Mathlib/Topology/EMetricSpace/PairReduction.lean

```diff
@@ -0,0 +1,427 @@
+    (logSizeBallSeq J hJ a c i).point ∉ (logSizeBallSeq J hJ a c (i + 1)).finset := by
+  simp [finset_logSizeBallSeq_add_one, logSizeBallStruct.smallBall]
+
+lemma finset_logSizeBallSeq_add_one_ssubset (hJ : J.Nonempty) (i : ℕ)
+    (h : (logSizeBallSeq J hJ a c i).finset.Nonempty) :
+    (logSizeBallSeq J hJ a c (i + 1)).finset ⊂ (logSizeBallSeq J hJ a c i).finset := by
+    apply ssubset_of_subset_not_subset
+    · simp [finset_logSizeBallSeq_add_one]
+    refine Set.not_subset.mpr ⟨(logSizeBallSeq J hJ a c i).point, ?_, ?_⟩
+    · exact point_mem_finset_logSizeBallSeq hJ i h
+    · exact point_notMem_finset_logSizeBallSeq_add_one hJ i
+
+lemma card_finset_logSizeBallSeq_add_one_lt (hJ : J.Nonempty) (i : ℕ)
+    (h : (logSizeBallSeq J hJ a c i).finset.Nonempty) :
+    #(logSizeBallSeq J hJ a c (i + 1)).finset < #(logSizeBallSeq J hJ a c i).finset := by
+  simp [Finset.card_lt_card, finset_logSizeBallSeq_add_one_ssubset hJ i h]
+
+lemma card_finset_logSizeBallSeq_le (hJ : J.Nonempty) (i : ℕ) :
+    #(logSizeBallSeq J hJ a c i).finset ≤ #J - i := by
+  induction i with
+  | zero => simp [finset_logSizeBallSeq_zero]
+  | succ i ih =>
+    by_cases h : (logSizeBallSeq J hJ a c i).finset.Nonempty
+    · have := card_finset_logSizeBallSeq_add_one_lt hJ i h
+      omega
+    apply le_trans <| Finset.card_le_card (finset_logSizeBallSeq_add_one_subset hJ i)
+    suffices #(logSizeBallSeq J hJ a c i).finset = 0 by simp [this]
+    rwa [← not_ne_iff, Finset.card_ne_zero.not]
+
+lemma card_finset_logSizeBallSeq_card_eq_zero (hJ : J.Nonempty) :
+    #(logSizeBallSeq J hJ a c #J).finset = 0 := by
+  rw [← Nat.le_zero, ← tsub_self #J]
+  exact card_finset_logSizeBallSeq_le hJ #J
+
+lemma disjoint_smallBall_logSizeBallSeq (hJ : J.Nonempty) {i j : ℕ} (hij : i ≠ j) :
+    Disjoint
+      ((logSizeBallSeq J hJ a c i).smallBall c) ((logSizeBallSeq J hJ a c j).smallBall c) := by
+  wlog h : i < j generalizing i j
+  · exact Disjoint.symm <| this hij.symm <| (ne_iff_lt_iff_le.mpr (not_lt.mp h)).mp hij.symm
+  apply Finset.disjoint_of_subset_right
+  · exact (Finset.filter_subset _ _).trans (antitone_logSizeBallSeq_add_one_subset hJ h)
+  simp [finset_logSizeBallSeq_add_one, Finset.disjoint_sdiff]
+
+/-- Given a log-size ball sequence `(Vᵢ, tᵢ, rᵢ)`, we define the pair set sequence by
+`Kᵢ = {tᵢ} × {x ∈ Vᵢ | dist(tᵢ, x) ≤ rᵢc}`. -/
+noncomputable
+def pairSetSeq (J : Finset T) (a c : ℝ≥0∞) (n : ℕ) : Finset (T × T) :=
+  if hJ : J.Nonempty then
+    Finset.product {(logSizeBallSeq J hJ a c n).point} ((logSizeBallSeq J hJ a c n).ball c)
+  else ∅
+
+/-- Given the pair set sequence Kᵢ we define the pair set `K` by `K = ⋃ i, Kᵢ`. -/
+noncomputable
+def pairSet (J : Finset T) (a c : ℝ≥0∞) : Finset (T × T) :=
+  Finset.biUnion (Finset.range #J) (pairSetSeq J a c)
+
+lemma pairSet_empty_eq_empty (a c : ℝ≥0∞) : pairSet (∅ : Finset T) a c = ∅ := rfl
+
+lemma pairSet_subset : pairSet J a c ⊆ J.product J := by
```

Comment:
```suggestion
lemma pairSet_subset : pairSet J a c ⊆ J ×ˢ J := by
```
Same thing later on in the file. It's defeq to what you wrote, so the proofs should still work, but it's our standard spelling.